samza-rest/src/main/java/org/apache/samza/rest/proxy/job/YarnRestJobStatusProvider.java

```diff
@@ -0,0 +1,164 @@
+import org.apache.samza.rest.resources.YarnJobResourceConfig;
+import org.codehaus.jackson.map.DeserializationConfig;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An implementation of the {@link JobStatusProvider} that retrieves
+ * the job status from the YARN REST api.
+ */
+public class YarnRestJobStatusProvider implements JobStatusProvider {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(YarnRestJobStatusProvider.class);
+  private final String _apiEndpoint;
+  private final HttpClient _httpClient;
+  private final ObjectMapper _objectMapper;
+
+  public YarnRestJobStatusProvider(JobsResourceConfig config) {
+    YarnJobResourceConfig yarnConfig = new YarnJobResourceConfig(config);
+    this._objectMapper = new ObjectMapper();
+    this._objectMapper.configure(DeserializationConfig.Feature.UNWRAP_ROOT_VALUE, true);
+    this._httpClient = new HttpClient();
+    this._apiEndpoint = String.format("http://%s/ws/v1/cluster/apps?states=RUNNING,NEW,NEW_SAVING,SUBMITTED,ACCEPTED",
+        yarnConfig.getYarnResourceManagerEndpoint());
+  }
+
+  private static final String JOB_NAME_ID_FORMAT = "%s_%s";
+
+  @Override
+  public void getJobStatuses(Collection<Job> jobs) throws IOException, InterruptedException {
+    if (jobs == null || jobs.isEmpty()) {
+      return;
+    }
+
+    // We will identify jobs returned by the YARN application states by their qualified names, so build a map
+    // to translate back from that name to the JobInfo we wish to populate. This avoids parsing/delimiter issues.
+    final Map<String, Job> qualifiedJobToInfo = new HashMap<>();
+    for(Job job : jobs) {
+      qualifiedJobToInfo.put(getQualifiedJobName(new JobInstance(job.getJobName(), job.getJobId())), job);
+    }
+
+    // Hit the rest endpoint and fetch the current status
+    try {
+      byte[] response = httpGet(_apiEndpoint);
+      ApplicationInfoWrapper appInfoWrapper = _objectMapper.readValue(response, ApplicationInfoWrapper.class);
+      for (ApplicationInfo appInfo:  appInfoWrapper.getApps()) {
+        String qualifiedName = appInfo.getName();
+        JobStatus samzaStatus = yarnStateToSamzaStatus(YarnApplicationState.valueOf(appInfo.getState().toUpperCase()));
+        Job job = qualifiedJobToInfo.get(qualifiedName);
+
+        // If job is null, it wasn't requested.  The default status is STOPPED because there could be many
+        // application attempts in that status. Only update the job status if it's not STOPPED.
+        if (job != null && (job.getStatusDetail() == null || samzaStatus != JobStatus.STOPPED)) {
+          job.setStatusDetail(appInfo.getState());
+          job.setStatus(samzaStatus);
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to retrieve node info.", e);
+    }
+  }
+
+  @Override
+  public Job getJobStatus(JobInstance jobInstance)
```

Comment:
This is potentially very inefficient. How often is this called?

Reply:
Is called if there is any uses cases for checking a single job's status. Currently only the start/stop in samza-rest uses it.